### DISCUSSION TITLE
Logging Source Code Generators

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#14] [General]
# Logging Source Code Generators

Since .NET6 we have the possibility to define an easy way of logging common statements:
Meet Compile-time logging source generators. This article will show why we have them and how to use them. Of course a smaller benchmark will also follow.

https://steven-giesel.com/blogPost/48697958-4aee-474a-8920-e266d1d7b8fa

**Comment by mehdihadeli** [comment]
Thanks, For this article,
I have one problem when I move the `public partial void LogCreateUser(User user, DateTime when);` to another partial class `UserCreationService`, in this case compiler can't infeer the `ILogger` field.
I think the compiler has a bug here right now.

**Reply by linkdotnet** [reply]
Hmm that is weird. For us in bUnit that approach works very well. Here is an example: https://github.com/bUnit-dev/bUnit/blob/AngleSharpRenderer/src/bunit.web/RenderingV2/BunitRenderer.Log.cs

**Reply by mehdihadeli** [reply]
Static one works (you mentioned link), but partial instance member doesn't work :|

This is a sample:
``` csharp
public partial class UserCreationService
{
    private ILogger _logger;

    public UserCreationService(ILogger logger)
    {
        _logger = logger;
    }

    public void LogCreateUserBad(User user, DateTime when)
    {
        // That is the "worst" approach as you lose structured logging
        // Helpers like JetBrains Resharper or Rider will even warn you
        _logger.LogInformation($"Creating user: {user} at {when}");
    }

    public void LogCreateUserTraditionally(User user, DateTime when)
    {
        _logger.LogInformation("Creating user: {User} at {When}", user, when);
    }

    [LoggerMessage(Message = "Creating user: {user} at {when}", Level = LogLevel.Information, EventId = 1)]
    public static partial void LogStaticVersion(ILogger logger, User user, DateTime when);

    public static void ExampleUsage()
    {
        var logger = LoggerFactory.Create(l => l.AddConsole()).CreateLogger(typeof(UserCreationService));
        var userService = new UserCreationService(logger);
        var user = new User("Steven", "Giesel");
        userService.LogCreateUser(user, DateTime.Now);
    }
}

public partial class UserCreationService
{
    // The function has to be partial as the source code generator has to provide the implementation
    // We can also set the LogLevel and an EventId
    // The source code generator will automatically look for an `ILogger` field to log its stuff
    [LoggerMessage(Message = "Creating user: {user} at {when}", Level = LogLevel.Information, EventId = 1)]
    public partial void LogCreateUser(User user, DateTime when);
}
```

**Reply by linkdotnet** [reply]
Ah okay. Thanks for the update. Maybe it would be worth to report this back to the dotnet team!?

**Reply by mehdihadeli** [reply]
Solution is moving `private ILogger _logger;` field to second partial class. it is Wierid :|

**Reply by mehdihadeli** [reply]
> Ah okay. Thanks for the update. Maybe it would be worth to report this back to the dotnet team!?

Yes, They have a bug here

**Reply by mehdihadeli** [reply]
I think there is an issue for that before https://github.com/dotnet/runtime/issues/61547

**Reply by linkdotnet** [reply]
Ah it's freshly merged (> 1 month ago). So it will be part of .net7.
Good to hear.

Thanks for the great input @mehdihadeli love it!

**Reply by mehdihadeli** [reply]
> I think, it fixed in .Net 7 :D

**Comment by ernest-morariu** [comment]
I think your benchmark compares very different things. 
You say that you created a NoLog ILoogger that doen't do the actual logging, but it doesn't do the formatting of message either, and the magic of the numbers your present here comes exactly for that reason.  
The function LogCreateUserBad does the formatting of the message as soon as you call it while the other two functions only forward some parameters to some "empty functions" (this is the reason they have no cost). 

If you modify your NoLog as:

    public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception? exception, Func<TState, Exception?, string> formatter)
    {
        string msg = formatter(state, exception);
    }

Note that this version still doesn't do the actual logging, but it does the formatting of the messge. With this version, all the three functions perform the formatting, and here is the new/correct benchmark:

|                     Method            |     Mean |    Error  |   StdDev |  Gen 0 | Allocated |
|-------------------------------|----------:|---------:|----------:|---------:|------------:|
|           LogCreateUserBad     | 502.1 ns |  4.53 ns |  4.23 ns | 0.1564   |     656 B    |
| LogCreateUserTraditionally | 632.5 ns  | 11.74 ns | 10.98 ns | 0.1717 |     720 B    |
|              LogCreateUser        | 570.6 ns |  7.93 ns |  7.41 ns | 0.1621    |     680 B   |

The advantage of the compile time generated functions stands only in the fact they check the IsEnabled before the actual formatting and logging, so, if the current logging level is not allowed, the formatting and the loggin is avoided. 
However, if IsEnabled returns True, the time cost is the same. 

Now there are other ways of implementing logging that makes the function LogCreateUserBad even faster and simplier than the compile time generated functions.

**Reply by linkdotnet** [reply]
Thanks for the comment. Your reasoning might be correct and I did indeed have a completely wrong and misleading setup.

According to Microsoft itself (https://learn.microsoft.com/en-us/dotnet/core/extensions/high-performance-logging) it is not just about `IsEnabled` or not. And indeed, if you check the generated code - at least the whole topic around boxing is "handled by the attribute".
